transport/http/middleware_host_prefix.go

```diff
@@ -0,0 +1,35 @@
+package http
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/awslabs/smithy-go/middleware"
+)
+
+// HostPrefixMiddleware provides a middleware to set a host
+// prefix on the request url
+type HostPrefixMiddleware struct {
+}
+
+// ID the identifier for the HostPrefixMiddleware
+func (m *HostPrefixMiddleware) ID() string { return "HostPrefixMiddleware" }
+
+// HandleBuild adds the host prefix to request url if set on request
+func (m *HostPrefixMiddleware) HandleBuild(
+	ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler,
+) (
+	out middleware.BuildOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown request type %T", req)
+	}
+
+	// Add the host prefix
+	if len(req.HostPrefix) != 0 && !IsEndpointHostPrefixDisabled(ctx) {
+		req.URL.Host = req.HostPrefix + req.URL.Host
+	}
```

Comment:
I think one thing we need to consider with this change, is if `HostPrefix` should be a member on the `smithyhttp.Request` at all. I'd venture to say it should not, since the correct endpoint cannot be determined without calling `smithyhttp.Requerst.Build`

Reply:
Can this skip the additional `Build` middleware and set `HostPrefix` in the serialize middleware since the new `IsEndpointHostPrefixDisabled` flag to disable behavior. It seems the additional Build middleware isn't really needed.

Reply:
Keeping it in serialize didn't seem a good choice as we lack middleware slots atm. 

Serialize steps are getting easily crowded and maintaining order is difficult. 

Keeping it in Build ensure this is ran after everything is serialized and customized. I decided to add it as one of the first middlewares in build step. I think we need a better story on what goes in serialize steps

Reply:
Understand the crowdyness of Serialize. I think thats the suggestion to remove HostPrefix all together.